proposals/0000-exportable-named-default.rst

```diff
@@ -0,0 +1,461 @@
+4. For any two imported ``default`` declarations for the same class where one subsumes the other, we ignore the
+   subsumed declaration.
+5. If a class has neither a local ``default`` declaration nor an imported ``default`` declaration that subsumes all
+   other imported ``default`` declarations for the class, the conflict between the imports is unresolvable. The effect
+   is to ignore all ``default`` declarations for the class, so that no declaration is in effect in the module. The
+   compiler may choose to emit a warning in this case, but no error would be triggered about the imports. Of course an
+   error may be triggered in the body of the module if it contains an actual ambiguous type for the class with the
+   conflicting imported defaults, as per the following section.
+
+Note that a ``default`` declaration that repeats a type name more than once is perfectly valid, and sometimes may
+be necessary to resolve coflicts. For example, a module that imports two conflicting defaults
+
+::
+
+   default C (Int, Bool)
+
+and
+   
+::
+
+   default C (Bool, Int)
+
+may use a local declaration
+
+::
+
+   default C (Int, Bool, Int)
+
+to override the imports. Because this declaration subsumes both imported defaults it will not trigger any compiler
+warning.
+
+   
+Rules for disambiguation at the use site
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The disambiguation rules are a conservative extension of the existing rules in Haskell 2010, which state that
+ambiguous type variable *v* is defaultable if:
+
+    - *v* appears only in constraints of the form *C* *v*, where *C* is a class, and
+
+    - at least one of these classes is a numeric class, (that is, ``Num`` or a subclass of ``Num``), and
+
+    - all of these classes are defined in the Prelude or a standard library.
+
+    Each defaultable variable is replaced by the first type in the default list that is an instance of all the
+    ambiguous variable’s classes. It is a static error if no such type is found.
+
+The new rules instead require only that 
+
+- *v* appears in at least one constraint of the form *C* *v*, where *C* is a class.
+
+Informally speaking, the type selected for defaulting is the first type from the ``default`` list for class *C* that
+satisfies all constraints on type variable *v*. If there are multiple *C*\ `i`:subscript: *v* constraints with
+competing ``default`` declarations, they have to resolve to the same type.
+
+To make the design more explicit, the following algorithm *can* be used for default resolution, but any other method
+that achieves the same effect can be substitued:
+
+Let *S* be the complete set of unsolved constraints, and initialize *S*\ `x`:subscript: to an empty set of constraints. For
+every *v* that is free in *S*:
+
+0. Define *B*\ `v`:subscript:, the subset of S consisting of all constraints whose only free type variable is *v*.
```

Comment:
What's the purpose of introducing `B_v`? It seems redundant with your previous definition (in https://github.com/ghc-proposals/ghc-proposals/pull/409/commits/d313f4f9a3caee343a50fe1f1e6a4292684b3add) of `C_v = { C_i v | C_i v \in S }`.

In other words, constraints of the form `C_i v` already have a single free type variable `v`.

Reply:
> constraints of the form C_i v already have a single free type variable v

Yes, but so do constraints of the form `C ~ v`, `C2 v Int`, etc. Introduction of the `B_v` set makes the algorithm more general. This change preserves (more) backward compatibility, see the example in my previous comment.

Reply:
Aha, I get it now. We introduce `B_v` not so we can use it in the definition of `C_v` (where `S` would work just as well), but so we can refer to it in Step 4. 

That makes sense, as does your example in https://github.com/ghc-proposals/ghc-proposals/pull/409#issuecomment-903302542, but it's pretty subtle. I think one more sentence in Step 0 foreshadowing the use in Step 4 would help a lot.

> `B_v` is the set of all constraints that must be satisfied by any defaulting decision for `v`.